### ISSUE TITLE
Bug in generic types for universal render target?

### ISSUE THREAD

**Issue by Alloyed** [ryansolid/dom-expressions#277]
https://github.com/ryansolid/dom-expressions/blob/main/packages/dom-expressions/src/universal.d.ts

Hi, I'm trying to implement a custom renderer that does not target the dom at all (!). As far as I can tell, to do this I need to 

A. replace JSX.Element with a variant that does not reference dom nodes ✅ 
B. call `createRenderer<MyCustomNode>`.

Doing this creates types with (roughly) this signature:
```tsx
type JSX.Element = MyCustomNode | JSX.Element[] | null | undefined; //... etc
const render: (code: () => MyCustomNode, node: MyCustomNode) => () => void
```
The problem here is that the component definition `() => MyCustomNode` appears to be different type to `JSX.Element`. The (DOM) definition of JSX.Element includes all sorts of things that aren't just HTMLElement, but when implementing my renderer I only wrote custom methods in terms of my custom dom node type, and not the array/null/other things.

What I naively would've expected is something like this signature:
```tsx
interface Renderer<NodeType, JsxElementType> {
    const render: (code: () => JsxElementType, node: NodeType) => () => void;
}
```

Is the type declaration wrong here, or should I be implementing my renderer to handle all those extra types, too?

**Comment by jpdutoit** [comment]
I believe it should be the following:
```
interface Renderer<NodeType> {
    const render: (code: () => UniversalJSXElement<NodeType>, node: NodeType) => () => void;
}

type UniversalJSXElement<NodeType> =
  | NodeType
  | Array<UniversalJSXElement<NodeType>> 
  | (string & {})
  | number
  | boolean
  | null
  | undefined
```

**Comment by jpdutoit** [comment]
Yeah, this is a real problem. I work around it by basically copy-pasting all those type definitions
```
import { For as UntypedFor,} from "solid-js"

export const For = UntypedFor as <T extends readonly any[], U extends JSX.Element>(props: {
  each: T | undefined | null | false
  fallback?: JSX.Element
  children: (item: T[number], index: Accessor<number>) => U
}) => JSX.Element
```

It works, but I end up constantly fighting the VSCode auto-importer to import the methods from the correct file. So, ideally the definitions should be updated to be NodeType agnostic. I think those definitions live in the main solid repo, so an issue there would make sense.

I think a wider discussion is needed here. Currently the developer experience in Typescript is not great.

**Comment by chiefcll** [comment]
Thanks for the info - we're working on adding Types to our project which is universal renderer... https://github.com/lightning-js/solid
@frank-weindel

**Comment by ryansolid** [comment]
This has been fixed in Solid 2.0. We np longer associate JSX types with the main library and instead have a SolidElement that can be merged into the custom renderers types.
